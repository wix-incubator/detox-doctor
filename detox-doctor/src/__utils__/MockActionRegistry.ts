import { kebabCase, snakeCase } from 'lodash';
import { SKIPPED } from '../constants';
import type { RuleRegistry, Rule } from '../types';

export class MockActionRegistry implements RuleRegistry {
  getRule = jest.fn();
  getRules = jest.fn();

  static aRule(comment: string): Rule {
    return {
      id: `rule_${snakeCase(comment)}`,
      alias: `rule-${kebabCase(comment)}`,
      description: `A ${comment} rule`,
      check: async () => ({ status: SKIPPED }),
    };
  }

  static twoRules() {
    const registry = new MockActionRegistry();
    registry.getRules.mockReturnValue([
      {
        id: 'rule1',
        alias: 'first-rule',
        description: 'First rule',
      },
      {
        id: 'rule2',
        alias: 'second-rule',
        description: 'Second rule',
      },
    ]);

    return registry;
  }
}
