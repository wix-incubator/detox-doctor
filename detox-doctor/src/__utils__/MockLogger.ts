import type { Logger } from '../types';

export class MockLogger implements Logger {
  public debug = jest.fn();
  public log = jest.fn();

  clear(level: 'debug' | 'log' = 'log'): void {
    this[level].mockClear();
  }

  dump(level: 'debug' | 'log' = 'log'): string {
    return this[level].mock.calls.map((c) => c.join(' ')).join('\n');
  }

  dumpJSON(level: 'debug' | 'log' = 'log'): object[] {
    return this[level].mock.calls.map((c) => {
      return JSON.parse(c[0]);
    });
  }
}
