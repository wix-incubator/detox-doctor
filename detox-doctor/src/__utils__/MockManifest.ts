import type { PackageManifest } from '../types';

export class MockManifest implements PackageManifest {
  deleteDependency = jest.fn();
  findDependency = jest.fn();
  getDependencyVersion = jest.fn();
  listDependencies = jest.fn();
  moveDependency = jest.fn();
  setDependency = jest.fn();
  updateDependency = jest.fn();
  getPackageManager = jest.fn();
  getRepositoryDirectory = jest.fn();
  hasWorkspaces = jest.fn();
  isDirty = jest.fn();
  get = jest.fn();

  constructor(public name = '', public version = '') {}
}
