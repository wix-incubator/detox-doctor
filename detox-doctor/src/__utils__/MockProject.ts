import { memoize } from 'lodash';
import type { Project } from '../types';
import type { MockManifest } from './MockManifest';

export class MockProject implements Project {
  rootDir = '';
  manifests: Record<string, MockManifest> = {};

  deleteFile = jest.fn();
  exec = jest.fn();
  hasDirectory = jest.fn();
  hasFile = jest.fn();
  readFile = jest.fn();
  writeFile = jest.fn();
  getManifest = jest.fn(memoize(async (name = '') => this.manifests[name] ?? null));
  installPackages = jest.fn();
  isModuleInstalled = jest.fn();
}
