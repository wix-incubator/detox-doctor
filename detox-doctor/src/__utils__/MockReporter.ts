import type { Reporter } from '../types';

export class MockReporter implements Reporter {
  static readonly instances: MockReporter[] = [];

  onRunStart = jest.fn();
  onRuleStart = jest.fn();
  onRuleResult = jest.fn();
  onRunEnd = jest.fn();

  constructor() {
    MockReporter.instances.push(this);
  }
}
