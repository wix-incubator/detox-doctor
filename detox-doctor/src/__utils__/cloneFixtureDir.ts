import path from 'path';
import fse from 'fs-extra';
import tempfile from 'tempfile';

declare const __TEMPORARY_FILES__: string[] | undefined;

export async function cloneFixtureDir(fixtureDirectory?: string): Promise<string> {
  const temporaryDirectory = tempfile('');
  await fse.ensureDir(path.dirname(temporaryDirectory));
  if (fixtureDirectory) {
    await fse.copy(fixtureDirectory, temporaryDirectory);
  }

  if (__TEMPORARY_FILES__ !== undefined) {
    __TEMPORARY_FILES__.push(temporaryDirectory);
  }

  return temporaryDirectory;
}
