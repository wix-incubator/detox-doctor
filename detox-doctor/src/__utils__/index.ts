export * from './MockActionRegistry';
export * from './MockLogger';
export * from './MockManifest';
export * from './MockProject';
export * from './MockReporter';
export * from './MockRule';

export * from './cloneFixtureDir';
export * from './instantiateRule';
