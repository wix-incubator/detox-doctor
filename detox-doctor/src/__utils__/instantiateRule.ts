import { ProjectImpl } from '../core';
import type { Rule, RuleConstructor } from '../types';
import { instantiateMaybeClass } from '../utils/instantiateMaybeClass';
import { cloneFixtureDir } from './cloneFixtureDir';
import { MockLogger } from './MockLogger';

export type InstantiateRule$Options = {
  bareMode?: boolean;
  cwd: string | undefined;
};

export async function instantiateRule<T extends Rule>(
  RuleClass: RuleConstructor<T>,
  options: InstantiateRule$Options,
) {
  const rootDir = await cloneFixtureDir(options.cwd);
  const logger = new MockLogger();
  const project = new ProjectImpl({ rootDir });
  const manifest = await project.getManifest();
  const rule = instantiateMaybeClass(RuleClass, {
    context: { bare: options.bareMode ?? false },
    logger,
    project,
    manifest,
  }) as T;
  return { rootDir, logger, project, manifest, rule };
}
