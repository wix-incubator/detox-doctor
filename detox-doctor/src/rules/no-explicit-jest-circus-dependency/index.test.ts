import { MockManifest } from '../../__utils__';
import { NoExplicitJestCircusDependencyRule } from './index';

describe('no-explicit-jest-circus-dependency', () => {
  let manifest: MockManifest;
  let rule: NoExplicitJestCircusDependencyRule;

  describe('#check', () => {
    beforeEach(() => {
      manifest = new MockManifest();
      rule = new NoExplicitJestCircusDependencyRule({ manifest });
    });

    test('no manifest', async () => {
      rule = new NoExplicitJestCircusDependencyRule({ manifest: null });
      await expect(rule.check()).resolves.toMatchSnapshot();
    });

    test('no jest', async () => {
      manifest.getDependencyVersion.mockReturnValue(void 0);
      manifest.findDependency.mockReturnValue([]);

      await expect(rule.check()).resolves.toMatchSnapshot();
    });

    test('non-semantic version', async () => {
      manifest.getDependencyVersion.mockReturnValue('latest');
      await expect(rule.check()).resolves.toMatchSnapshot();
    });

    test('old version', async () => {
      manifest.getDependencyVersion.mockReturnValue('^26.6.0');
      await expect(rule.check()).resolves.toMatchSnapshot();
    });

    test('has circus', async () => {
      manifest.getDependencyVersion.mockReturnValue('^27.2.5');
      manifest.findDependency.mockReturnValue(['devDependencies']);
      await expect(rule.check()).resolves.toMatchSnapshot();
    });

    test('has no circus', async () => {
      manifest.getDependencyVersion.mockReturnValue('^27.2.5');
      manifest.findDependency.mockReturnValue([]);

      await expect(rule.check()).resolves.toMatchSnapshot();
    });
  });

  describe('#fix', () => {
    beforeEach(() => {
      manifest = new MockManifest();
      rule = new NoExplicitJestCircusDependencyRule({ manifest });
    });

    test('happy path', async () => {
      manifest.getDependencyVersion.mockReturnValue('^27.2.5');
      manifest.deleteDependency.mockReturnValue(true);
      await expect(rule.fix()).resolves.toMatchSnapshot();
    });

    test('unhappy path', async () => {
      manifest.getDependencyVersion.mockReturnValue(void 0);
      manifest.deleteDependency.mockReturnValue(false);
      await expect(rule.fix()).resolves.toMatchSnapshot();
    });
  });
});
