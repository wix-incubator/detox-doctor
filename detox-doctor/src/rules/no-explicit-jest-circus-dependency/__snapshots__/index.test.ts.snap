// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no-explicit-jest-circus-dependency #check has circus 1`] = `
{
  "message": "Project should not have jest-circus as a dependency",
  "status": "failure",
}
`;

exports[`no-explicit-jest-circus-dependency #check has no circus 1`] = `
{
  "message": undefined,
  "status": "success",
}
`;

exports[`no-explicit-jest-circus-dependency #check no jest 1`] = `
{
  "message": undefined,
  "status": "success",
}
`;

exports[`no-explicit-jest-circus-dependency #check no manifest 1`] = `
{
  "message": "Couldn't find package.json in your project.",
  "status": "skipped",
}
`;

exports[`no-explicit-jest-circus-dependency #check non-semantic version 1`] = `
{
  "message": "The project is using non-semver version of jest: latest",
  "status": "skipped",
}
`;

exports[`no-explicit-jest-circus-dependency #check old version 1`] = `
{
  "message": "The project is using jest@^26.6.0, which is not affected by this rule",
  "status": "skipped",
}
`;

exports[`no-explicit-jest-circus-dependency #fix happy path 1`] = `
{
  "message": "Removed jest-circus@^27.2.5 from dependencies",
  "status": "success",
}
`;

exports[`no-explicit-jest-circus-dependency #fix unhappy path 1`] = `
{
  "message": "Failed to remove jest-circus@undefined from dependencies",
  "status": "failure",
}
`;
